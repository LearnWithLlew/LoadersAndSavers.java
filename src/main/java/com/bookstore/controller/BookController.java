package com.bookstore.controller;

import com.bookstore.model.*;
import com.spun.util.persistence.Loader;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Controller
public class BookController {

    public final DataSource dataSource;

    @Autowired
    public BookController(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    @GetMapping("/")
    public String listBooks(Model model) {
        List<Book> books1 = new ArrayList<>();
        Map<Long, Book> bookMap = new HashMap<>();
        Map<Long, Publisher> publisherMap = new HashMap<>();
        Map<Long, Address> addressMap = new HashMap<>();
        Map<Long, City> cityMap = new HashMap<>();
        Map<Long, State> stateMap = new HashMap<>();
        Map<Long, Country> countryMap = new HashMap<>();

        String sql = "SELECT b.*, p.*, a.*, c.id as city_id, c.name as city_name, c.postal_code, " +
                "s.id as state_id, s.name as state_name, s.code as state_code, " +
                "co.id as country_id, co.name as country_name, co.code as country_code " +
                "FROM Books b " +
                "LEFT JOIN Publishers p ON p.id = b.publisher_id " +
                "LEFT JOIN Addresses a ON a.id = p.address_id " +
                "LEFT JOIN Cities c ON c.id = a.city_id " +
                "LEFT JOIN States s ON s.id = c.state_id " +
                "LEFT JOIN Countries co ON co.id = s.country_id " +
                "ORDER BY b.id FETCH FIRST 10 ROWS ONLY";

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {

            while (rs.next()) {
                try {
                    // Process Book
                    Long bookId = rs.getLong("ID");
                    Book book = bookMap.get(bookId);
                    if (book == null) {
                        book = new Book();
                        book.setId(bookId);
                        book.setTitle(rs.getString("TITLE"));
                        book.setSubtitle(rs.getString("SUBTITLE"));
                        book.setIsbn13(rs.getString("ISBN13"));
                        book.setIsbn10(rs.getString("ISBN10"));
                        book.setEdition(rs.getString("EDITION"));
                        book.setPageCount(rs.getInt("PAGE_COUNT"));
                        book.setLanguage(rs.getString("LANGUAGE"));
                        book.setDescription(rs.getString("DESCRIPTION"));
                        book.setCoverImageUrl(rs.getString("COVER_IMAGE_URL"));
                        book.setPrice(rs.getBigDecimal("PRICE"));

                        java.sql.Date pubDate = rs.getDate("PUBLICATION_DATE");
                        if (pubDate != null) {
                            book.setPublicationDate(pubDate.toLocalDate());
                        }

                        // Create default objects to avoid null pointer exceptions in the template
                        Publisher publisher = new Publisher();
                        Address address = new Address();
                        City city = new City();
                        State state = new State();

                        // Set up the object graph with default objects
                        city.setState(state);
                        address.setCity(city);
                        publisher.setAddress(address);
                        book.setPublisher(publisher);

                        bookMap.put(bookId, book);
                        books1.add(book);
                    }

                    // Process Publisher
                    Long publisherId = rs.getLong("PUBLISHER_ID");
                    if (!rs.wasNull()) {
                        Publisher publisher = publisherMap.get(publisherId);
                        if (publisher == null) {
                            publisher = new Publisher();
                            publisher.setId(publisherId);
                            publisher.setName(rs.getString("NAME"));
                            publisher.setWebsite(rs.getString("WEBSITE"));
                            publisher.setEmail(rs.getString("EMAIL"));
                            publisher.setPhone(rs.getString("PHONE"));

                            // Create default objects to avoid null pointer exceptions
                            Address address = new Address();
                            City city = new City();
                            State state = new State();

                            // Set up the object graph with default objects
                            city.setState(state);
                            address.setCity(city);
                            publisher.setAddress(address);

                            publisherMap.put(publisherId, publisher);
                        }
                        book.setPublisher(publisher);

                        // Process Address
                        Long addressId = rs.getLong("ADDRESS_ID");
                        if (!rs.wasNull()) {
                            Address address = addressMap.get(addressId);
                            if (address == null) {
                                address = new Address();
                                address.setId(addressId);
                                address.setStreetLine1(rs.getString("STREET_LINE1"));
                                address.setStreetLine2(rs.getString("STREET_LINE2"));

                                // Create default objects to avoid null pointer exceptions
                                City city = new City();
                                State state = new State();

                                // Set up the object graph with default objects
                                city.setState(state);
                                address.setCity(city);

                                addressMap.put(addressId, address);
                            }
                            publisher.setAddress(address);

                            // Process City
                            Long cityId = rs.getLong("CITY_ID");
                            if (!rs.wasNull()) {
                                City city = cityMap.get(cityId);
                                if (city == null) {
                                    city = new City();
                                    city.setId(cityId);
                                    city.setName(rs.getString("city_name"));
                                    city.setPostalCode(rs.getString("POSTAL_CODE"));

                                    // Create default objects to avoid null pointer exceptions
                                    State state = new State();

                                    // Set up the object graph with default objects
                                    city.setState(state);

                                    cityMap.put(cityId, city);
                                }
                                address.setCity(city);

                                // Process State
                                Long stateId = rs.getLong("STATE_ID");
                                if (!rs.wasNull()) {
                                    State state = stateMap.get(stateId);
                                    if (state == null) {
                                        state = new State();
                                        state.setId(stateId);
                                        state.setName(rs.getString("state_name"));
                                        state.setCode(rs.getString("state_code"));
                                        stateMap.put(stateId, state);
                                    }
                                    city.setState(state);

                                    // Process Country
                                    Country country = new Country(rs);
                                    state.setCountry(country);
                                    countryMap.put(country.getId(), country);
                                }
                            }
                        }
                    }
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        List<Book> books = books1;
        model.addAttribute("books", books);
        return "index";
    }

    @GetMapping("/V2")
    public String listBooksV2(Model model) {
        return listBooks(model, null);
    }

    public static String listBooks(Model model, Loader<List<Book>> loader) {
        model.addAttribute("books", loader.load());
        return "index";
    }
}
