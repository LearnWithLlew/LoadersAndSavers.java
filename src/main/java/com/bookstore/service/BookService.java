package com.bookstore.service;

import com.bookstore.model.Book;
import com.bookstore.repository.BookRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;


@Service
public class BookService {

    private final BookRepository bookRepository;

    @Autowired
    public BookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    public List<Book> getTop10Books() {
        Pageable pageable = PageRequest.of(0, 10);
        return bookRepository.findTop10Books(pageable);
    }
}
