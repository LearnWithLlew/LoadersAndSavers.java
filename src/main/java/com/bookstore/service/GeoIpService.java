package com.bookstore.service;

import com.maxmind.geoip2.DatabaseReader;
import com.maxmind.geoip2.exception.GeoIp2Exception;
import com.maxmind.geoip2.model.CountryResponse;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.InetAddress;

@Service
public class GeoIpService {
    private final DatabaseReader dbReader;

    public GeoIpService() {
        try {
            InputStream database = getClass().getClassLoader().getResourceAsStream("GeoLite2-Country.mmdb");
            if (database == null) {
                throw new IOException("GeoLite2-Country.mmdb not found in classpath");
            }
            File tempFile = File.createTempFile("GeoLite2-Country", ".mmdb");
            tempFile.deleteOnExit();
            try (FileOutputStream out = new FileOutputStream(tempFile)) {
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = database.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            }
            this.dbReader = new DatabaseReader.Builder(tempFile).build();
        } catch (IOException e) {
            throw new RuntimeException("Failed to load GeoLite2-Country.mmdb", e);
        }
    }

    public String getCountryName(String ip) {
        try {
            InetAddress ipAddress = InetAddress.getByName(ip);
            CountryResponse response = dbReader.country(ipAddress);
            if (response.getCountry() != null) {
                return response.getCountry().getName();
            }
        } catch (IOException | GeoIp2Exception e) {
            // log error or handle as needed
        }
        return null;
    }
}
