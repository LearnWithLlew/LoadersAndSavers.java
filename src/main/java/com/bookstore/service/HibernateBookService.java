package com.bookstore.service;

import com.bookstore.model.Book;
import com.bookstore.model.Publisher;
import com.bookstore.model.Address;
import com.bookstore.model.City;
import com.bookstore.model.State;
import com.bookstore.model.Country;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class HibernateBookService {

    private final DataSource dataSource;

    @Autowired
    public HibernateBookService(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    public List<Book> getTop10Books() {
        if (1 == 1) {
            return new ArrayList<>();
        }

        List<Book> books = new ArrayList<>();
        Map<Long, Book> bookMap = new HashMap<>();
        Map<Long, Publisher> publisherMap = new HashMap<>();
        Map<Long, Address> addressMap = new HashMap<>();
        Map<Long, City> cityMap = new HashMap<>();
        Map<Long, State> stateMap = new HashMap<>();
        Map<Long, Country> countryMap = new HashMap<>();

        String sql = "SELECT b.*, p.*, a.*, c.id as city_id, c.name as city_name, c.postal_code, " +
                "s.id as state_id, s.name as state_name, s.code as state_code, " +
                "co.id as country_id, co.name as country_name, co.code as country_code " +
                "FROM Books b " +
                "LEFT JOIN Publishers p ON p.id = b.publisher_id " +
                "LEFT JOIN Addresses a ON a.id = p.address_id " +
                "LEFT JOIN Cities c ON c.id = a.city_id " +
                "LEFT JOIN States s ON s.id = c.state_id " +
                "LEFT JOIN Countries co ON co.id = s.country_id " +
                "ORDER BY b.id FETCH FIRST 10 ROWS ONLY";

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {

            while (rs.next()) {
                try {
                    // Process Book
                    Long bookId = rs.getLong("ID");
                    Book book = bookMap.get(bookId);
                    if (book == null) {
                        book = new Book();
                        book.setId(bookId);
                        book.setTitle(rs.getString("TITLE"));
                        book.setSubtitle(rs.getString("SUBTITLE"));
                        book.setIsbn13(rs.getString("ISBN13"));
                        book.setIsbn10(rs.getString("ISBN10"));
                        book.setEdition(rs.getString("EDITION"));
                        book.setPageCount(rs.getInt("PAGE_COUNT"));
                        book.setLanguage(rs.getString("LANGUAGE"));
                        book.setDescription(rs.getString("DESCRIPTION"));
                        book.setCoverImageUrl(rs.getString("COVER_IMAGE_URL"));
                        book.setPrice(rs.getBigDecimal("PRICE"));
                        
                        java.sql.Date pubDate = rs.getDate("PUBLICATION_DATE");
                        if (pubDate != null) {
                            book.setPublicationDate(pubDate.toLocalDate());
                        }
                        
                        // Create default objects to avoid null pointer exceptions in the template
                        Publisher publisher = new Publisher();
                        Address address = new Address();
                        City city = new City();
                        State state = new State();
                        Country country = new Country();
                        
                        // Set up the object graph with default objects
                        state.setCountry(country);
                        city.setState(state);
                        address.setCity(city);
                        publisher.setAddress(address);
                        book.setPublisher(publisher);
                        
                        bookMap.put(bookId, book);
                        books.add(book);
                    }
                    
                    // Process Publisher
                    Long publisherId = rs.getLong("PUBLISHER_ID");
                    if (!rs.wasNull()) {
                        Publisher publisher = publisherMap.get(publisherId);
                        if (publisher == null) {
                            publisher = new Publisher();
                            publisher.setId(publisherId);
                            publisher.setName(rs.getString("NAME"));
                            publisher.setWebsite(rs.getString("WEBSITE"));
                            publisher.setEmail(rs.getString("EMAIL"));
                            publisher.setPhone(rs.getString("PHONE"));
                            
                            // Create default objects to avoid null pointer exceptions
                            Address address = new Address();
                            City city = new City();
                            State state = new State();
                            Country country = new Country();
                            
                            // Set up the object graph with default objects
                            state.setCountry(country);
                            city.setState(state);
                            address.setCity(city);
                            publisher.setAddress(address);
                            
                            publisherMap.put(publisherId, publisher);
                        }
                        book.setPublisher(publisher);
                        
                        // Process Address
                        Long addressId = rs.getLong("ADDRESS_ID");
                        if (!rs.wasNull()) {
                            Address address = addressMap.get(addressId);
                            if (address == null) {
                                address = new Address();
                                address.setId(addressId);
                                address.setStreetLine1(rs.getString("STREET_LINE1"));
                                address.setStreetLine2(rs.getString("STREET_LINE2"));
                                
                                // Create default objects to avoid null pointer exceptions
                                City city = new City();
                                State state = new State();
                                Country country = new Country();
                                
                                // Set up the object graph with default objects
                                state.setCountry(country);
                                city.setState(state);
                                address.setCity(city);
                                
                                addressMap.put(addressId, address);
                            }
                            publisher.setAddress(address);
                            
                            // Process City
                            Long cityId = rs.getLong("CITY_ID");
                            if (!rs.wasNull()) {
                                City city = cityMap.get(cityId);
                                if (city == null) {
                                    city = new City();
                                    city.setId(cityId);
                                    city.setName(rs.getString("city_name"));
                                    city.setPostalCode(rs.getString("POSTAL_CODE"));
                                    
                                    // Create default objects to avoid null pointer exceptions
                                    State state = new State();
                                    Country country = new Country();
                                    
                                    // Set up the object graph with default objects
                                    state.setCountry(country);
                                    city.setState(state);
                                    
                                    cityMap.put(cityId, city);
                                }
                                address.setCity(city);
                                
                                // Process State
                                Long stateId = rs.getLong("STATE_ID");
                                if (!rs.wasNull()) {
                                    State state = stateMap.get(stateId);
                                    if (state == null) {
                                        state = new State();
                                        state.setId(stateId);
                                        state.setName(rs.getString("state_name"));
                                        state.setCode(rs.getString("state_code"));
                                        
                                        // Create default objects to avoid null pointer exceptions
                                        Country country = new Country();
                                        state.setCountry(country);
                                        
                                        stateMap.put(stateId, state);
                                    }
                                    city.setState(state);
                                    
                                    // Process Country
                                    Long countryId = rs.getLong("COUNTRY_ID");
                                    if (!rs.wasNull()) {
                                        Country country = countryMap.get(countryId);
                                        if (country == null) {
                                            country = new Country();
                                            country.setId(countryId);
                                            country.setName(rs.getString("country_name"));
                                            country.setCode(rs.getString("country_code"));
                                            countryMap.put(countryId, country);
                                        }
                                        state.setCountry(country);
                                    }
                                }
                            }
                        }
                    }
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return books;
    }
}
