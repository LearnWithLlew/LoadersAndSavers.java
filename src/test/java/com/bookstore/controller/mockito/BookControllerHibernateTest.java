package com.bookstore.controller.mockito;

import com.bookstore.controller.BookController;
import com.bookstore.controller.ThymeleafUtils;
import com.bookstore.service.HibernateBookService;
import org.apache.derby.jdbc.EmbeddedDataSource;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.FileLauncherReporter;
import org.approvaltests.reporters.MultiReporter;
import org.junit.jupiter.api.Test;
import org.springframework.core.io.ClassPathResource;
import org.springframework.jdbc.datasource.init.ResourceDatabasePopulator;
import org.springframework.ui.ConcurrentModel;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class BookControllerHibernateTest {

    @Test
    public void testDirectRenderingOfThymeleafTemplate() throws Exception {
        HibernateBookService hibernateBookService = new HibernateBookService(getDataSource());

        var model = new ConcurrentModel();

        String page = new BookController(null, hibernateBookService).listBooksV3(model);
        String htmlOutput = ThymeleafUtils.renderPage(page, model);

        Approvals.verifyHtml(htmlOutput,
            new Options().withReporter(new MultiReporter(DiffReporter.INSTANCE, new FileLauncherReporter())));
    }

    public DataSource getDataSource() throws SQLException {
        var dataSource = mock(DataSource.class);
        var connection = mock(Connection.class);
        when(dataSource.getConnection()).thenReturn(connection);

        var statement = mock(java.sql.PreparedStatement.class);
        var resultSet = mock(java.sql.ResultSet.class);
        when(connection.prepareStatement(org.mockito.ArgumentMatchers.anyString())).thenReturn(statement);
        when(statement.executeQuery()).thenReturn(resultSet);

        when(resultSet.next()).thenReturn(true, false);
        when(resultSet.getLong(org.mockito.ArgumentMatchers.anyInt())).thenReturn(1L);
        when(resultSet.getString(org.mockito.ArgumentMatchers.anyInt())).thenReturn("Title");
        when(resultSet.getInt(org.mockito.ArgumentMatchers.anyInt())).thenReturn(2024);

        return dataSource;
    }

}
