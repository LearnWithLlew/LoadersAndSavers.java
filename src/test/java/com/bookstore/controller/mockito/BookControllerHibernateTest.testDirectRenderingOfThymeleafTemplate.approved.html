<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Bookstore - Top 10 Books</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            line-height: 1.6;
            margin: 0;
            padding: 20px;
            color: #333;
        }
        h1 {
            color: #2c3e50;
            text-align: center;
            margin-bottom: 30px;
        }
        .book-list {
            display: grid;
            grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
            gap: 20px;
            max-width: 1200px;
            margin: 0 auto;
        }
        .book-card {
            border: 1px solid #ddd;
            border-radius: 8px;
            padding: 15px;
            box-shadow: 0 2px 5px rgba(0,0,0,0.1);
            transition: transform 0.3s ease;
        }
        .book-card:hover {
            transform: translateY(-5px);
        }
        .book-title {
            font-size: 1.2rem;
            font-weight: bold;
            margin-bottom: 10px;
            color: #2c3e50;
        }
        .book-subtitle {
            font-size: 1rem;
            color: #7f8c8d;
            margin-bottom: 10px;
        }
        .book-details {
            font-size: 0.9rem;
            margin-bottom: 5px;
        }
        .book-price {
            font-weight: bold;
            color: #e74c3c;
            font-size: 1.1rem;
            margin-top: 10px;
        }
        .no-books {
            text-align: center;
            font-size: 1.2rem;
            color: #7f8c8d;
            margin-top: 50px;
        }
    </style>
</head>
<body>
<h1>Top 10 Books</h1>



<div class="book-list">
    <div class="book-card">
        <div class="book-title">Twilit</div>
        <div class="book-subtitle">A Vampire Romance Parody</div>
        <div class="book-subtitle">Girl falls in love with a slightly illuminated vampire</div>
        <div class="book-details">ISBN-13: <span>9783456789012</span></div>
        <div class="book-details">Published: <span>05 Oct 2008</span></div>
        <div class="book-details">Pages: <span>498</span></div>
        <div class="book-details">Language: <span>English</span></div>
        <div class="book-details">Country: <span>United States</span></div>
        <div class="book-price">$12.99</div>
    </div>
</div>
</body>
</html>
